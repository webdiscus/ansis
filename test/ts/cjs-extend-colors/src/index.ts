import ansis, { Ansis, AnsiColors } from 'ansis';

type AnsiColorsExtend<T extends string> = AnsiColors | (T & Record<never, never>);

const log = console.log;

// npm package for CSS color names: https://chir.ag/projects/name-that-color/
// tools to get name by hex: https://www.npmjs.com/package/css-color-names
const myTheme = {
  pink: '#FF75D1',
  orange: '#FFAB40',
  apple: '#4FA83D',
};

// Define the function using the extracted valid styles from the extended instance
const myLog = (style: AnsiColorsExtend<keyof typeof myTheme>, message: string) => {
  console.log(ansis[style](message));
};

// `AnsiColorsExtend` is an extendable type to add a custom color in the default color list
const myLog2 = (style: AnsiColorsExtend<'pink' | 'apple'>, message: string) => {
  console.log(ansis[style](message));
};

// Extend base colors
const color = ansis.extend(myTheme);

// Access extended colors
const { pink, orange, red } = color;

myLog('red', 'default: red'); // default style, OK
myLog('cyanBright', 'default: cyanBright'); // default style, OK
myLog('pink', 'extended: pink'); // extended style, OK
myLog('orange', 'extended: orange'); // extended style, OK
myLog2('apple', 'extended: apple'); // OK

// Expected: TS2345: Argument of type 'unknown' is not assignable to parameter of type AnsiColorsExtend<'pink' | 'orange' | 'apple'>
//myLog('unknown', 'message');

//log(customAnsis.apple.italic`extended color: apple italic`);
log(color.apple.italic`extended color: apple italic`);
log(red`destructured default color: red`);
log(pink`destructured extended color: pink`);
log(orange.italic`destructured extended color: orange italic`);

log(color.pink.underline('extended chained color: pink underline'));
log(pink.underline('destructured extended chained color: pink underline'));

// test in TS both extended variants: normal and background color names
log(color.orange('Orange foreground'));
log(color.bgOrange('Orange background'));

// Usage test: This should simulate what you'd get in TS validation

function testExtend(inst: Ansis) {
  const themed = inst.extend({
    foo: '#ff0',
    bar: { open: '\x1b[31m', close: '\x1b[39m' },
  });

  log(themed.foo('should be yellow'));
  log(themed.bar('should be red'));
}

function testExtend2(inst: InstanceType<typeof Ansis>) {
  const themed = inst.extend({ baz: '#ff0000' });

  log(themed.baz('should be red'));
}

testExtend(ansis);
testExtend2(ansis);