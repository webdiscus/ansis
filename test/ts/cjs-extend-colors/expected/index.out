[31mdefault: red[39m
[96mdefault: cyanBright[39m
[38;2;255;117;209mextended: pink[39m
[38;2;255;171;64mextended: orange[39m
[38;2;79;168;61mextended: apple[39m
[38;2;79;168;61m[3mextended color: apple italic[23m[39m
[31mdestructured default color: red[39m
[38;2;255;117;209mdestructured extended color: pink[39m
[38;2;255;171;64m[3mdestructured extended color: orange italic[23m[39m
[38;2;255;117;209m[4mextended chained color: pink underline[24m[39m
[38;2;255;117;209m[4mdestructured extended chained color: pink underline[24m[39m
[38;2;255;171;64mOrange foreground[39m
[48;2;255;171;64mOrange background[49m
[38;2;255;255;0mshould be yellow[39m
[31mshould be red[39m
[38;2;255;0;0mshould be red[39m
